Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,348 @@
+"""Testing `tabnanny` module.
+
+
+Glossary:
+    * errored    : Whitespace related problems present in file.
+"""
+from unittest import TestCase, mock
+from unittest.mock import patch
+import tabnanny
+import tokenize
+from tempfile import NamedTemporaryFile, TemporaryDirectory
+from test.support import (captured_stderr, captured_stdout, script_helper,
+                          findfile)
+
+
+SOURCE_CODES = {
+    "incomplete_expression": (
+        'fruits = [\n'
+        '    "Apple",\n'
+        '    "Orange",\n'
+        '    "Banana",\n'
+        '\n'
+        'print(fruits)\n'
+    ),
+    "wrong_indented": (
+        'if True:\n'
+        '    print("hello")\n'
+        '  print("world")\n'
+        'else:\n'
+        '    print("else called")\n'
+    ),
+    "nannynag_errored": (
+        'if True:\n'
+        ' \tprint("hello")\n'
+        '\tprint("world")\n'
+        'else:\n'
+        '    print("else called")\n'
+    ),
+    "error_free": (
+        'if True:\n'
+        '    print("hello")\n'
+        '    print("world")\n'
+        'else:\n'
+        '    print("else called")\n'
+    ),
+    "tab_space_errored_1": (
+        'def my_func():\n'
+        '\t  print("hello world")\n'
+        '\t  if True:\n'
+        '\t\tprint("If called")'
+    ),
+    "tab_space_errored_2": (
+        'def my_func():\n'
+        '\t\tprint("Hello world")\n'
+        '\t\tif True:\n'
+        '\t        print("If called")'
+    )
+}
+
+
+def create_tmp_python_file(source_code, dir=None):
+    """Common utility function for writing source code to temporary file.
+
+    It creates file with `.py` extension.
+
+    Returns: Absolute path of created code file."""
+    with NamedTemporaryFile(mode='w', dir=dir,
+                            delete=False, suffix=".py") as f:
+        f.write(source_code)
+    return f.name
+
+
+class TestFormatWitnesses(TestCase):
+    """Testing `tabnanny.format_witnesses()`."""
+
+    def test_format_whitenesses(self):
+        """Asserting formatter result by giving various input samples."""
+        tests = [
+            {'words': 'Test' , 'expected': 'at tab sizes T, e, s, t'},
+            {'words': ''     , 'expected': 'at tab size '},
+            {'words': 't'    , 'expected': 'at tab size t'},
+            {'words': '  t  ', 'expected': 'at tab sizes  ,  , t,  ,  '},
+        ]
+
+        for test in tests:
+            self.assertEqual(
+                tabnanny.format_witnesses(test['words']), test['expected']
+            )
+
+class TestErrPrint(TestCase):
+    """Testing `tabnanny.errprint()`."""
+
+    def test_errprint(self):
+        """Asserting result of `tabnanny.errprint()` by giving sample inputs."""
+        tests = [
+            {'args': ['first', 'second'], 'expected': 'first second\n'},
```

Comment:
Don't think we should kill the readability by converting it to `tuple`. How about `collections.namedtuple`?